### PR TITLE
feat: fix resource(work/resource in member clusters) conflicting between PP and MCS.

### PR DESCRIPTION
**Pull request by jwcesign**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:
Previously, we directly generated service work, but this led to conflicts with work created based on RB/PP. To resolve this issue, we introduced a new naming style for the work.

To prevent these conflicts, we propose a solution: MCS will generate RB from the service and if there's an existing RB generated by PP, MCS will assume control over it. This way, no conflict arises. Additionally, once a resource is matched by MCS, the detector should ignore it.

Looking ahead, other CRs similar to MCS(like a scenario-related PP) may emerge in future scenarios. They should adhere to this approach to avoid potential conflicts.

**Which issue(s) this PR fixes**:
Fixes #none

**Special notes for your reviewer**:
none

**Does this PR introduce a user-facing change?**:

```release-note
`karmada-controller-manager`: Fix resource(work/resource in member clusters) conflicting between PP and MCS.
```

### CODE REVIEW

File: pkg/apis/networking/v1alpha1/service_types.go [hunk 1]
```diff
@@ -79,20 +79,40 @@ type MultiClusterServiceSpec struct {
...
 
 	// ServiceConsumptionClusters specifies the clusters where the service will be exposed, for clients.
 	// If leave it empty, the service will be exposed to all clusters.
+	// Deprecated: in favor of ProviderClusters/ConsumerClusters.
 	// +optional
 	ServiceConsumptionClusters []string `json:"serviceConsumptionClusters,omitempty"`
+
+	// ProviderClusters specifies the clusters which will provision the service backend.
+	// If leave it empty, we will collect the backend endpoints from all clusters and sync
+	// them to the ServiceConsumptionClusters.
```
Comment (RainbowMango) [inline pkg/apis/networking/v1alpha1/service_types.go:101]:
```suggestion
	// them to the ConsumerClusters.
```

File: pkg/apis/networking/v1alpha1/service_types.go [hunk 2]
```diff
@@ -79,20 +79,27 @@ type MultiClusterServiceSpec struct {
...
 	// +optional
 	Range ExposureRange `json:"range,omitempty"`
 
-	// ServiceProvisionClusters specifies the clusters which will provision the service backend.
+	// ProviderClusters specifies the clusters which will provision the service backend.
 	// If leave it empty, we will collect the backend endpoints from all clusters and sync
 	// them to the ServiceConsumptionClusters.
 	// +optional
-	ServiceProvisionClusters []string `json:"serviceProvisionClusters,omitempty"`
+	ProviderClusters []ClusterSelector `json:"providerClusters,omitempty"`
```
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/apis/networking/v1alpha1/service_types.go:103]: Looks good.

File: pkg/apis/networking/v1alpha1/well_known_constants.go [hunk 3]
```diff
@@ -23,4 +23,10 @@ const (
 	// The reason for generating a new unique identifier instead of simply using metadata.UUID is because:
 	// In backup scenarios, when applying the backup resource manifest in a new cluster, the UUID may change.
 	MultiClusterServicePermanentIDLabel = "multiclusterservice.karmada.io/permanent-id"
+
+	// MultiClusterServiceNameAnnotation is the name of a MultiClusterService object.
+	MultiClusterServiceNameAnnotation = "multiclusterservice.karmada.io/name"
+
+	// MultiClusterServiceNamespaceAnnotation is the namespace of a MultiClusterService object.
+	MultiClusterServiceNamespaceAnnotation = "multiclusterservice.karmada.io/namespace"
```
Comment (RainbowMango) [inline pkg/apis/networking/v1alpha1/well_known_constants.go:31]: Unused?

File: pkg/apis/work/v1alpha2/well_known_constants.go [hunk 4]
```diff
@@ -82,6 +82,9 @@ const (
 
 	// WorkNameLabel is added to objects to specify associated Work's name.
 	WorkNameLabel = "work.karmada.io/name"
+
+	// WorkClusterNameLabel is added to ResoruceBinding to represent what kind of resource manage this Binding.
+	BindingManagedByLabel = "binding.karmada.io/managed-by"
```
Comment (RainbowMango) [inline pkg/apis/work/v1alpha2/well_known_constants.go:87]: Unused and comments mismatch.

File: pkg/apis/networking/v1alpha1/well_known_constants.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/apis/networking/v1alpha1/well_known_constants.go:31]: Ignore me...

Comment (jwcesign) [reply pkg/apis/networking/v1alpha1/well_known_constants.go:31]: it's used in mcs_controller.go

File: pkg/util/constants.go [hunk 5]
```diff
@@ -58,6 +58,8 @@ const (
 	// resourcetemplate.karmada.io/retain-replicas: true   // with value `true` indicates retain
 	// resourcetemplate.karmada.io/retain-replicas: false  // with value `false` and others, indicates not retain
 	RetainReplicasLabel = "resourcetemplate.karmada.io/retain-replicas"
+
+	ResourceTemplateMatchedByLabel = "resorucetemplate.karmada.io/matched-by"
```
Comment (RainbowMango) [inline pkg/util/constants.go:62]: How do you think the:
```suggestion
	ResourceTemplateMatchedByLabel = "resorucetemplate.karmada.io/claimed-by"
```

File: pkg/detector/detector.go [hunk 6]
```diff
@@ -247,6 +247,12 @@ func (d *ResourceDetector) Reconcile(key util.QueueKey) error {
 		return err
 	}
 
+	resoruceTemplateMatchedBy := util.GetLabelValue(object.GetLabels(), util.ResourceTemplateMatchedByLabel)
```
Comment (RainbowMango) [inline pkg/detector/detector.go:250]:
```suggestion
	resourceTemplateMatchedBy := util.GetLabelValue(object.GetLabels(), util.ResourceTemplateMatchedByLabel)
```

File: pkg/detector/detector.go [hunk 7]
```diff
@@ -247,6 +247,12 @@ func (d *ResourceDetector) Reconcile(key util.QueueKey) error {
...
+	if resoruceTemplateMatchedBy != "" && resoruceTemplateMatchedBy != util.PolicyKind {
```
Comment (RainbowMango) [inline pkg/detector/detector.go:251]:
```suggestion
	if resoruceTemplateMatchedBy != ""  {
```

File: pkg/events/events.go [hunk 8]
```diff
@@ -132,12 +132,14 @@ const (
...
 // Define events for MultiClusterService objects with CrossCluster type.
 const (
-	// EventReasonSyncServiceWorkFailed is indicates that sync service work failed.
-	EventReasonSyncServiceWorkFailed string = "SyncServiceWorkFailed"
-	// EventReasonSyncServiceWorkSucceed is indicates that sync service work succeed.
-	EventReasonSyncServiceWorkSucceed string = "SyncServiceWorkSucceed"
+	// EventReasonSyncServiceFailed is indicates that sync service work failed.
+	EventReasonSyncServiceFailed string = "SyncServiceWorkFailed"
+	// EventReasonSyncServiceSucceed is indicates that sync service work succeed.
+	EventReasonSyncServiceSucceed string = "SyncServiceWorkSucceed"
```
Comment (RainbowMango) [inline pkg/events/events.go:138]:
```suggestion
	EventReasonSyncServiceFailed string = "SyncServiceFailed"
	// EventReasonSyncServiceSucceed is indicates that sync service work succeed.
	EventReasonSyncServiceSucceed string = "SyncServiceSucceed"
```

File: pkg/events/events.go [hunk 9]
```diff
@@ -132,12 +132,14 @@ const (
...
 	// EventReasonDispatchEndpointSliceFailed indicates that dispatch endpointslice failed.
 	EventReasonDispatchEndpointSliceFailed = "DispatchEndpointSliceFailed"
 	// EventReasonDispatchEndpointSliceSucceed indicates that dispatch endpointslice succeed.
 	EventReasonDispatchEndpointSliceSucceed = "DispatchEndpointSliceSucceed"
+	// EventReasonConfigurationRedundant indicates that MultiClusterService configuration redundant.
+	EventReasonConfigurationRedundant = "ConfigurationRedundant"
```
Comment (RainbowMango) [inline pkg/events/events.go:144]: consider another name.

File: pkg/apis/work/v1alpha2/well_known_constants.go [hunk 10]
```diff
@@ -82,6 +82,9 @@ const (
 
 	// WorkNameLabel is added to objects to specify associated Work's name.
 	WorkNameLabel = "work.karmada.io/name"
+
+	// BindingManagedByLabel is added to ResoruceBinding to represent what kind of resource manages this Binding.
```
Comment (RainbowMango) [inline pkg/apis/work/v1alpha2/well_known_constants.go:86]:
```suggestion
	// BindingManagedByLabel is added to ResourceBinding to represent what kind of resource manages this Binding.
```

File: pkg/scheduler/event_handler.go [hunk 11]
```diff
@@ -108,7 +108,8 @@ func (s *Scheduler) resourceBindingEventFilter(obj interface{}) bool {
 	}
 
 	return util.GetLabelValue(accessor.GetLabels(), policyv1alpha1.PropagationPolicyNameLabel) != "" ||
-		util.GetLabelValue(accessor.GetLabels(), policyv1alpha1.ClusterPropagationPolicyLabel) != ""
+		util.GetLabelValue(accessor.GetLabels(), policyv1alpha1.ClusterPropagationPolicyLabel) != "" ||
+		util.GetLabelValue(accessor.GetLabels(), workv1alpha2.BindingManagedByLabel) != ""
```
Comment (RainbowMango) [inline pkg/scheduler/event_handler.go:112]: TODO: consider a new label name.

File: pkg/controllers/multiclusterservice/endpointslice_dispatch_controller.go [hunk 12]
```diff
@@ -284,7 +283,7 @@ func (c *EndpointsliceDispatchController) cleanOrphanDispatchedEndpointSlice(ctx
 			continue
 		}
 
-		consumptionClusters, err := helper.GetConsumptionClustres(c.Client, mcs)
+		consumptionClusters, _, err := helper.GetConsumerClustres(c.Client, mcs)
```
Comment (jwcesign) [inline pkg/controllers/multiclusterservice/endpointslice_dispatch_controller.go:286]: consumerClusters

File: pkg/controllers/multiclusterservice/mcs_controller.go [hunk 13]
```diff
@@ -85,44 +83,60 @@ func (c *MCSController) Reconcile(ctx context.Context, req controllerruntime.Req
...
-	if err = c.handleMCSCreateOrUpdate(ctx, mcs.DeepCopy()); err != nil {
+	if err = c.handleMultiClusterServiceCreateOrUpdate(mcs.DeepCopy(), providerClusters, consumerClusters); err != nil {
 		return controllerruntime.Result{}, err
 	}
 	return controllerruntime.Result{}, nil
 }
 
-func (c *MCSController) handleMCSDelete(ctx context.Context, mcs *networkingv1alpha1.MultiClusterService) (controllerruntime.Result, error) {
+func (c *MCSController) handleMultiClusterServiceDelete(mcs *networkingv1alpha1.MultiClusterService) (controllerruntime.Result, error) {
 	klog.V(4).InfoS("Begin to handle MultiClusterService delete event", "namespace", mcs.Namespace, "name", mcs.Name)
```
Comment (jwcesign) [inline pkg/controllers/multiclusterservice/mcs_controller.go:123]: Make the logging format unified

File: pkg/controllers/multiclusterservice/mcs_controller.go [hunk 14]
```diff
@@ -209,7 +183,7 @@ func (c *MCSController) deleteMultiClusterServiceWork(mcs *networkingv1alpha1.Mu
 		}
 	}
 
-	klog.V(4).InfoS("Success to delete MultiClusterService work", "namespace", mcs.Namespace, "name", mcs.Name)
+	klog.V(4).InfoS("Success to delete MultiClusterService(%s/%s) work:%v", mcs.Namespace, mcs.Name, err)
```
Comment (jwcesign) [inline pkg/controllers/multiclusterservice/mcs_controller.go:186]: Make the logging format unified

File: pkg/controllers/multiclusterservice/mcs_controller.go [hunk 15]
```diff
@@ -364,68 +317,194 @@ func (c *MCSController) ensureMultiClusterServiceWork(ctx context.Context, mcs *
...
+			OwnerReferences: []metav1.OwnerReference{
+				{APIVersion: svc.APIVersion, Kind: svc.Kind, Name: svc.Name, UID: svc.UID},
+			},
+			Annotations: map[string]string{
+				networkingv1alpha1.MultiClusterServiceNameAnnotation:      mcs.Name,
+				networkingv1alpha1.MultiClusterServiceNamespaceAnnotation: mcs.Namespace,
+			},
 			Labels: map[string]string{
-				// We add this id in mutating webhook, let's just use it
+				workv1alpha2.BindingManagedByLabel:                     util.MultiClusterServiceKind,
```
Comment (jwcesign) [inline pkg/controllers/multiclusterservice/mcs_controller.go:450]: Use controller name

File: pkg/util/constants.go [hunk 16]
```diff
@@ -172,6 +175,10 @@ const (
 	ServiceExportKind = "ServiceExport"
 	// ServiceImportKind indicates the target resource is a serviceimport crd
 	ServiceImportKind = "ServiceImport"
+
+	PolicyKind = "Policy"
```
Comment (RainbowMango) [inline pkg/util/constants.go:179]: Unused.

File: pkg/controllers/multiclusterservice/mcs_controller.go [hunk 17]
```diff
@@ -502,6 +581,7 @@ func (c *MCSController) SetupWithManager(mgr controllerruntime.Manager) error {
 
 			// We only care about the update events below:
 			if equality.Semantic.DeepEqual(svcOld.Annotations, svcNew.Annotations) &&
+				equality.Semantic.DeepEqual(svcOld.Labels, svcNew.Labels) &&
```
Comment (jwcesign) [inline pkg/controllers/multiclusterservice/mcs_controller.go:584]: It's better to claim that we do not care about status updating

File: pkg/util/constants.go [hunk 18]
```diff
@@ -172,6 +175,10 @@ const (
...
+
+	MultiClusterServiceKind = "MultiClusterService"
```
Comment (jwcesign) [inline pkg/util/constants.go:181]: Add comment

File: pkg/util/constants.go (hunk 16 above)
Comment (RainbowMango) [reply pkg/util/constants.go:179]: Generally looks good now.

File: pkg/controllers/multiclusterservice/mcs_controller.go (hunk 15 above)
Comment (jwcesign) [reply pkg/controllers/multiclusterservice/mcs_controller.go:449]: Ignore this, it's better to use ResourceKind

File: pkg/util/helper/mcs.go [hunk 19]
```diff
@@ -106,28 +106,34 @@ func MultiClusterServiceCrossClusterEnabled(mcs *networkingv1alpha1.MultiCluster
...
 	}
-	return provisionClusters.Intersection(existClusters), nil
+	return provisionClusters.Clone().Intersection(allClusters), provisionClusters.Clone().Delete(allClusters.UnsortedList()...), nil
 }
 
-func GetConsumptionClustres(client client.Client, mcs *networkingv1alpha1.MultiClusterService) (sets.Set[string], error) {
-	consumptionClusters := sets.New[string](mcs.Spec.ServiceConsumptionClusters...)
-	existClusters, err := util.GetClusterSet(client)
+func GetConsumerClustres(client client.Client, mcs *networkingv1alpha1.MultiClusterService) (existClusters, noneExistClusters sets.Set[string], err error) {
+	consumptionClusters := sets.New[string]()
```
Comment (jwcesign) [inline pkg/util/helper/mcs.go:126]: consumerClusters

File: test/e2e/mcs_test.go [hunk 20]
```diff
@@ -748,21 +752,30 @@ var _ = ginkgo.Describe("CrossCluster MultiClusterService testing", func() {
 			})
 
 			gomega.Eventually(func() bool {
-				return checkEndpointSliceWithMultiClusterService(testNamespace, mcsName, mcs.Spec.ServiceProvisionClusters, mcs.Spec.ServiceConsumptionClusters)
+				return checkEndpointSliceWithMultiClusterService(testNamespace, mcsName, mcs.Spec.ProviderClusters, mcs.Spec.ConsumerClusters)
```
Comment (jwcesign) [inline test/e2e/mcs_test.go:755]: Please update the Consumption to Consumer, Provision to Provider in this e2e file.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**